File: packages/hydris/src/container/DiContainer.ts
import { type Binding, binding } from './Binding';
import {
    CircularDependencyError,
    ForbiddenDependencyError,
    ScopeRequiredError,
    UnboundTokenError,
} from './errors';
import { describe, type TokenLike } from './Token';
import {
    type BuildableLifetime,
    type Disposable,
    type Factory,
    type Service,
    type ServiceClass,
    DiServiceType,
} from './types';

const SERVICE_LIFETIME: Record<DiServiceType, BuildableLifetime> = {
    [DiServiceType.Singleton]: 'singleton',
    [DiServiceType.Scoped]: 'scoped',
    [DiServiceType.Transient]: 'transient',
};

const lifetimeOf = (declared?: DiServiceType): BuildableLifetime =>
    SERVICE_LIFETIME[declared ?? DiServiceType.Singleton];

type Registration = {
    lifetime: BuildableLifetime;
    factory: Factory<unknown>;
};

const hasDispose = (value: unknown): value is Disposable =>
    typeof value === 'object' &&
    value !== null &&
    typeof (value as { dispose?: unknown }).dispose === 'function';

export class DiContainer {
    private registrations: Map<object, Registration> | null = null;
    private instances: Map<object, unknown> | null = null;
    private disposables: Disposable[] | null = null;
    private building: Set<object> | null = null;
    private forbidden: Set<string> | null = null;
    private contextByToken: Map<object, string> | null = null;
    private contextHint: Map<string, string> | null = null;
    private readonly root_: DiContainer;

    constructor(private readonly parent: DiContainer | null = null) {
        this.root_ = parent === null ? this : parent.root_;
    }

    value<T>(token: TokenLike<T>, instance: T): this {
        this.instanceMap().set(token as object, instance);
        return this;
    }

    singleton<T>(token: TokenLike<T>, factory: Factory<T>): this {
        this.register(token as object, { lifetime: 'singleton', factory });
        return this;
    }

    scoped<T>(token: TokenLike<T>, factory: Factory<T>): this {
        this.register(token as object, { lifetime: 'scoped', factory });
        return this;
    }

    transient<T>(token: TokenLike<T>, factory: Factory<T>): this {
        this.register(token as object, { lifetime: 'transient', factory });
        return this;
    }

    bind(cls: ServiceClass): Binding {
        const factory = autoFactory(cls);
        const key = cls as object;
        const set = (lifetime: BuildableLifetime): void => {
            this.register(key, { lifetime, factory });
        };
        set('singleton');
        return binding(set);
    }

    has(token: TokenLike<unknown>): boolean {
        const key = token as object;
        for (let c: DiContainer | null = this; c !== null; c = c.parent) {
            if (c.instances?.has(key) || c.registrations?.has(key)) return true;
        }
        return false;
    }

    start(services: Service[] = []): this {
        for (const service of services) this.registerService(service);
        if (this.registrations !== null) {
            for (const [token, registration] of this.registrations) {
                if (registration.lifetime === 'singleton') {
                    this.get(token as TokenLike<unknown>);
                }
            }
        }
        return this;
    }

    private registerService(service: Service): void {
        if (typeof service === 'function') {
            this.register(service as object, {
                lifetime: lifetimeOf(service.service),
                factory: autoFactory(service),
            });
            return;
        }
        if ('value' in service) {
            this.value(service.provide, service.value);
            return;
        }
        const provide = service.provide;
        const declared =
            typeof provide === 'function'
                ? (provide as ServiceClass).service
                : undefined;
        this.register(provide as object, {
            lifetime: lifetimeOf(declared),
            factory: service.factory,
        });
    }

    scope(): DiContainer {
        return new Scope(this);
    }

    contextTokens(
        name: string,
        tokens: readonly TokenLike<unknown>[],
        hint?: string,
    ): this {
        const root = this.root_;
        if (root.contextByToken === null) root.contextByToken = new Map();
        for (const t of tokens) root.contextByToken.set(t as object, name);
        if (hint !== undefined) {
            if (root.contextHint === null) root.contextHint = new Map();
            root.contextHint.set(name, hint);
        }
        return this;
    }

    forbid(name: string): this {
        if (this.forbidden === null) this.forbidden = new Set();
        this.forbidden.add(name);
        return this;
    }

    noHttp(): this {
        return this.forbid('http');
    }

    private forbids(context: string): boolean {
        for (let c: DiContainer | null = this; c !== null; c = c.parent) {
            if (c.forbidden?.has(context)) return true;
        }
        return false;
    }

    get<T>(token: TokenLike<T>): T {
        return this.resolve(token as object) as T;
    }

    dispose(): void | Promise<void> {
        this.instances?.clear();
        const disposables = this.disposables;
        if (disposables === null || disposables.length === 0) return;
        return this.disposeAll(disposables);
    }

    private async disposeAll(disposables: Disposable[]): Promise<void> {
        for (let i = disposables.length - 1; i >= 0; i--) {
            await disposables[i].dispose();
        }
        disposables.length = 0;
    }

    private resolve(key: object): unknown {
        const context = this.root_.contextByToken?.get(key);
        if (context !== undefined && this.forbids(context)) {
            const building = this.root_.building;
            const chain = building === null ? [key] : [...building, key];
            throw new ForbiddenDependencyError(
                context,
                chain.map(describe),
                this.root_.contextHint?.get(context),
            );
        }

        for (let c: DiContainer | null = this; c !== null; c = c.parent) {
            const instances = c.instances;
            if (instances?.has(key)) {
                return instances.get(key);
            }
        }

        let registration: Registration | undefined;
        for (let c: DiContainer | null = this; c !== null; c = c.parent) {
            registration = c.registrations?.get(key);
            if (registration !== undefined) break;
        }

        if (registration === undefined) {
            if (typeof key === 'function') {
                return this.runFactory(
                    key,
                    autoFactory(key as ServiceClass),
                    this,
                );
            }
            throw new UnboundTokenError(describe(key));
        }

        switch (registration.lifetime) {
            case 'singleton': {
                const root = this.root_;
                const instance = this.runFactory(
                    key,
                    registration.factory,
                    root,
                );
                root.instanceMap().set(key, instance);
                root.remember(instance);
                return instance;
            }
            case 'scoped': {
                if (this.parent === null) {
                    throw new ScopeRequiredError(describe(key));
                }
                const instance = this.runFactory(
                    key,
                    registration.factory,
                    this,
                );
                this.instanceMap().set(key, instance);
                this.remember(instance);
                return instance;
            }
            default:
                return this.runFactory(key, registration.factory, this);
        }
    }

    private runFactory(
        key: object,
        factory: Factory<unknown>,
        resolver: DiContainer,
    ): unknown {
        const building = this.root_.buildingSet();
        if (building.has(key)) {
            throw new CircularDependencyError([...building, key].map(describe));
        }
        building.add(key);
        try {
            return factory(resolver);
        } finally {
            building.delete(key);
        }
    }

    private remember(instance: unknown): void {
        if (hasDispose(instance)) this.disposableList().push(instance);
    }

    private register(key: object, registration: Registration): void {
        this.registrationMap().set(key, registration);
    }

    private registrationMap(): Map<object, Registration> {
        if (this.registrations === null) this.registrations = new Map();
        return this.registrations;
    }

    private instanceMap(): Map<object, unknown> {
        if (this.instances === null) this.instances = new Map();
        return this.instances;
    }

    private disposableList(): Disposable[] {
        if (this.disposables === null) this.disposables = [];
        return this.disposables;
    }

    private buildingSet(): Set<object> {
        if (this.building === null) this.building = new Set();
        return this.building;
    }
}

class Scope extends DiContainer {
    override singleton(): this {
        return this.denied();
    }

    override scoped(): this {
        return this.denied();
    }

    override transient(): this {
        return this.denied();
    }

    override bind(): Binding {
        return this.denied();
    }

    override start(): this {
        return this.denied();
    }

    private denied(): never {
        throw new Error(
            'Cannot register services on a request scope; register them on the root container.',
        );
    }
}

const INJECT = Symbol.for('ssr.inject');

const autoFactory =
    (cls: ServiceClass): Factory<object> =>
    (c: DiContainer): object => {
        const deps = (cls as { [INJECT]?: readonly unknown[] })[INJECT] ?? [];
        const args = deps.map((dep) => c.get(dep as TokenLike<unknown>));
        const build = cls as unknown as new (...args: unknown[]) => object;
        return new build(...args);
    };

File: packages/hydris-sharp/Hydris/Container/DiContainer.cs
using System.Collections.Concurrent;
using System.Collections.Frozen;

namespace Hydris.Container;

public sealed class DiContainer : IAsyncDisposable {
    private readonly record struct Registration(Lifetime Lifetime, Func<DiContainer, object> Factory);

    private readonly DiContainer? Parent;
    private readonly DiContainer Root;
    private readonly ConcurrentDictionary<Type, object>? Shared;
    private readonly object? BuildLock;

    private Dictionary<Type, object>? Local;
    private Dictionary<Type, Registration>? Registrations;
    private FrozenDictionary<Type, Registration>? FrozenRegistrations;
    private List<object>? Disposables;
    private HashSet<string>? Forbidden;
    private Dictionary<Type, string>? ContextByToken;
    private Dictionary<string, string>? ContextHints;

    [ThreadStatic]
    private static List<Type>? BuildChain;

    public DiContainer() {
        Parent = null;
        Root = this;
        Shared = new ConcurrentDictionary<Type, object>();
        BuildLock = new object();
    }

    private DiContainer(DiContainer parent) {
        Parent = parent;
        Root = parent.Root;
    }

    public DiContainer Value<T>(T instance) {
        ArgumentNullException.ThrowIfNull(instance);
        SetLocal(typeof(T), instance);
        return this;
    }

    public DiContainer Singleton<T>(Func<DiContainer, T> factory) => Register<T>(Lifetime.Singleton, factory);

    public DiContainer Scoped<T>(Func<DiContainer, T> factory) => Register<T>(Lifetime.Scoped, factory);

    public DiContainer Transient<T>(Func<DiContainer, T> factory) => Register<T>(Lifetime.Transient, factory);

    private DiContainer Register<T>(Lifetime lifetime, Func<DiContainer, T> factory) {
        ArgumentNullException.ThrowIfNull(factory);
        if (Parent is not null)
            throw new InvalidOperationException("Cannot register services on a scope; register them on the root container.");

        (Registrations ??= [])[typeof(T)] = new Registration(lifetime, container => factory(container)!);
        FrozenRegistrations = null;
        return this;
    }

    public T Get<T>() => (T)Resolve(typeof(T));

    public object Get(Type type) {
        ArgumentNullException.ThrowIfNull(type);
        return Resolve(type);
    }

    public bool Has<T>() => Has(typeof(T));

    public bool Has(Type type) {
        ArgumentNullException.ThrowIfNull(type);
        for (DiContainer? c = this; c is not null; c = c.Parent) {
            if (c.TryGetLocal(type, out _))
                return true;
            if (c.TryGetRegistration(type, out _))
                return true;
        }

        return false;
    }

    public DiContainer Scope() => new(this);

    public DiContainer Start() {
        if (Parent is not null)
            throw new InvalidOperationException("Start applies to the root container.");

        if (Registrations is null)
            return this;

        FrozenRegistrations = Registrations.ToFrozenDictionary();
        foreach (var (key, registration) in Registrations) {
            if (registration.Lifetime == Lifetime.Singleton)
                Resolve(key);
        }

        return this;
    }

    public DiContainer ContextTokens(string name, IEnumerable<Type> tokens, string? hint = null) {
        var root = Root;
        root.ContextByToken ??= [];
        foreach (var token in tokens)
            root.ContextByToken[token] = name;

        if (hint is not null) {
            root.ContextHints ??= [];
            root.ContextHints[name] = hint;
        }

        return this;
    }

    public DiContainer Forbid(string name) {
        Forbidden ??= [];
        Forbidden.Add(name);
        return this;
    }

    public DiContainer NoHttp() => Forbid("http");

    public async ValueTask DisposeAsync() {
        Shared?.Clear();
        Local?.Clear();

        if (Disposables is not null) {
            for (var i = Disposables.Count - 1; i >= 0; i--) {
                switch (Disposables[i]) {
                    case IAsyncDisposable async:
                        await async.DisposeAsync();
                        break;
                    case IDisposable sync:
                        sync.Dispose();
                        break;
                }
            }

            Disposables.Clear();
        }
    }

    private object Resolve(Type key) {
        GuardContext(key);

        for (DiContainer? c = this; c is not null; c = c.Parent) {
            if (c.TryGetLocal(key, out var cached))
                return cached;
        }

        Registration? found = null;
        for (DiContainer? c = this; c is not null; c = c.Parent) {
            if (c.TryGetRegistration(key, out var registration)) {
                found = registration;
                break;
            }
        }

        if (found is null)
            throw new UnboundTokenException(Describe(key));

        var reg = found.Value;
        return reg.Lifetime switch {
            Lifetime.Singleton => Root.BuildCached(key, reg.Factory),
            Lifetime.Scoped => BuildScoped(key, reg.Factory),
            _ => Build(key, reg.Factory, this),
        };
    }

    private object BuildScoped(Type key, Func<DiContainer, object> factory) {
        if (Parent is null)
            throw new ScopeRequiredException(Describe(key));
        return BuildCached(key, factory);
    }

    private object BuildCached(Type key, Func<DiContainer, object> factory) {
        if (TryGetLocal(key, out var existing))
            return existing;

        if (Shared is null) {
            var scoped = Build(key, factory, this);
            SetLocal(key, scoped);
            Remember(scoped);
            return scoped;
        }

        lock (BuildLock!) {
            if (TryGetLocal(key, out existing))
                return existing;

            var instance = Build(key, factory, this);
            SetLocal(key, instance);
            Remember(instance);
            return instance;
        }
    }

    private object Build(Type key, Func<DiContainer, object> factory, DiContainer resolver) {
        var chain = BuildChain ??= [];
        if (chain.Contains(key)) {
            var cycle = new List<string>(chain.Count + 1);
            foreach (var link in chain)
                cycle.Add(Describe(link));
            cycle.Add(Describe(key));
            throw new CircularDependencyException(cycle);
        }

        chain.Add(key);
        try {
            return factory(resolver);
        } finally {
            chain.RemoveAt(chain.Count - 1);
        }
    }

    private void Remember(object instance) {
        if (instance is IAsyncDisposable or IDisposable)
            (Disposables ??= []).Add(instance);
    }

    private void GuardContext(Type key) {
        var context = Root.ContextByToken?.GetValueOrDefault(key);
        if (context is null || !Forbids(context))
            return;

        var chain = BuildChain;
        var names = new List<string>((chain?.Count ?? 0) + 1);
        if (chain is not null) {
            foreach (var link in chain)
                names.Add(Describe(link));
        }

        names.Add(Describe(key));
        throw new ForbiddenDependencyException(context, names, Root.ContextHints?.GetValueOrDefault(context));
    }

    private bool Forbids(string context) {
        for (DiContainer? c = this; c is not null; c = c.Parent) {
            if (c.Forbidden?.Contains(context) == true)
                return true;
        }

        return false;
    }

    private bool TryGetLocal(Type key, out object value) {
        if (Shared is not null)
            return Shared.TryGetValue(key, out value!);
        if (Local is not null)
            return Local.TryGetValue(key, out value!);
        value = null!;
        return false;
    }

    private void SetLocal(Type key, object value) {
        if (Shared is not null)
            Shared[key] = value;
        else
            (Local ??= [])[key] = value;
    }

    private bool TryGetRegistration(Type key, out Registration registration) {
        var frozen = FrozenRegistrations;
        if (frozen is not null)
            return frozen.TryGetValue(key, out registration);
        if (Registrations is not null)
            return Registrations.TryGetValue(key, out registration);
        registration = default;
        return false;
    }

    private static string Describe(Type type) => type.Name;
}
